tensorflow/compiler/mlir/tensorflow/transforms/convert_tf_control_flow_to_scf.cc

```diff
@@ -0,0 +1,117 @@
+      rewriter.eraseBlock(&scfThenOrElseRegion.back());
+
+      Block* firstBlockOfScfThenOrElseRegion = &scfThenOrElseRegion.front();
+
+      // Replace the current terminator (a `tf.Yield` op) with an `scf.yield`
+      // op. The input of the `scf.yield` op is a list of results of `tf.Cast`
+      // ops, each of which casts an operand of the current terminator to the
+      // corresponding result type of the `tf.IfRegion` op.
+      Operation* currentTerminator =
+          firstBlockOfScfThenOrElseRegion->getTerminator();
+      rewriter.setInsertionPoint(currentTerminator);
+      SmallVector<Value, 4> scfYieldInput;
+      for (auto it :
+           llvm::zip(tfIfRegionReturnType, currentTerminator->getOperands())) {
+        scfYieldInput.push_back(rewriter.create<CastOp>(
+            currentTerminator->getLoc(), std::get<0>(it), std::get<1>(it)));
+      }
+      auto scfYield = rewriter.create<scf::YieldOp>(currentTerminator->getLoc(),
+                                                    scfYieldInput);
+      rewriter.eraseOp(currentTerminator);
+    };
+
+    Location loc = op.getLoc();
+
+    // The condition of an `scf.if` op is a 1-bit signless integer. Whereas, the
+    // condition of the `tf.IfRegion` op is a 0-D tensor of 1-bit signless
+    // integers. Thus, we use the `tensor.extract` op to compute the condition
+    // of `scf.if` from that of `tf.IfRegion`.
+    auto scfIfCondition = rewriter.create<tensor::ExtractOp>(loc, op.cond());
+
+    TypeRange tfIfRegionReturnType = op.getResultTypes();
+
+    // Create the `scf.if` op.
+    auto scfIfOp = rewriter.create<scf::IfOp>(
+        loc, tfIfRegionReturnType, scfIfCondition, /*withElseRegion=*/true);
+
+    Region& thenRegion = op.then_branch();
+    Region& elseRegion = op.else_branch();
+
+    // Create the `then` and `else` regions of the `scf.if` op.
+    createScfThenOrElse(thenRegion, scfIfOp.thenRegion(), tfIfRegionReturnType,
+                        rewriter);
+    createScfThenOrElse(elseRegion, scfIfOp.elseRegion(), tfIfRegionReturnType,
+                        rewriter);
+
+    // Replace the `tf.IfRegion` op with the results of the `scf.if` op.
+    rewriter.replaceOp(op, scfIfOp.getResults());
+    return success();
+  }
+};
+
+}  // end anonymous namespace
+
+void populateTfControlFlowToScfPatterns(MLIRContext* context,
+                                        OwningRewritePatternList* patterns) {
+  patterns->insert<ConvertIfRegionOp>(context);
+}
+
+struct ConvertTfControlFlowToScf
+    : public PassWrapper<ConvertTfControlFlowToScf, OperationPass<ModuleOp>> {
+  llvm::StringRef getArgument() const final {
+    return "convert-tf-control-flow-to-scf";
+  }
+
+  llvm::StringRef getDescription() const final {
+    return "Convert TensorFlow control flow to SCF";
+  }
+
+  void getDependentDialects(DialectRegistry& registry) const override {
+    registry.insert<scf::SCFDialect, tensor::TensorDialect>();
```

Comment:
Because I am using the `tensor.extract` op to convert the `tensor<i1>` condition of `tf.IfRegion` to the `i1` condition of `scf.if`. An `scf.if` op's condition is of type `i1` while that of a `tf.IfRegion` op is `tensor<i1>`.

Reply:
Ohh... No problem ! :)